### PR TITLE
Add middlewares

### PR DESCRIPTION
**Pull request by urmastalimaa**

### CODE REVIEW

File: lib/salestation/web/request_logger.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+          query:        env[QUERY_STRING],
+          content_type: env[CONTENT_TYPE],
+          http_agent:   env[HTTP_USER_AGENT],
+          http_accept:  env[HTTP_ACCEPT],
+          server_name:  env[SERVER_NAME]
+        }
+      end
+
+      def log_response(env, status, headers, body, began_at)
+        type = status >= 500 ? :info : :error
```
Comment (willmore) [inline lib/salestation/web/request_logger.rb]: `type` not used until line 48; scope larger than needed; move to line ~47.

File: lib/salestation/web/request_logger.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        begin
+          # Rack body is a string
+          JSON.parse(body[0])
+        rescue Exception
+          {error: "Failed to parse response body"}
+        end
+      end
+
+      class Logger
+        def initialize(logger, request_id)
```
Comment (willmore) [inline lib/salestation/web/request_logger.rb]: Why is Logger stateful?

File: lib/salestation/web/request_logger.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+          )
+        })
+      end
+
+      private
+
+      def parse_body(body)
+        begin
+          # Rack body is a string
+          JSON.parse(body[0])
```
Comment (deiwin) [inline lib/salestation/web/request_logger.rb]: why the `0`?

File: lib/salestation/web/request_logger.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+          http_accept:  env[HTTP_ACCEPT],
+          server_name:  env[SERVER_NAME]
+        }
+      end
+
+      def log_response(env, status, headers, body, began_at)
+        type = status >= 500 ? :info : :error
+        response_object = parse_body(body)
+
+        if status >= 400
```
Comment (deiwin) [inline lib/salestation/web/request_logger.rb]: Why do we need to log twice if it's `>= 400`? Did you considered logging once, but using different data depending on whether it's below or above 400?

File: lib/salestation/web/request_logger.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            "Unsuccessful request",
+            error: response_object,
+            status: status.to_s
+          )
+        end
+
+        @logger.public_send(type, 'Processed request', {
+          path: env[REQUEST_URI],
+          method: env[REQUEST_METHOD],
+          status: status,
```
Comment (deiwin) [inline lib/salestation/web/request_logger.rb]: You use `status.to_s` above, but no `to_s` here.

File: lib/salestation/web/statsd_middleware.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+module Salestation
+  class Web < Module
+    class StatsdMiddleware
+
+      def initialize(app, statsd, metric:)
```
Comment (deiwin) [inline lib/salestation/web/statsd_middleware.rb]: Where is this initialized?

File: lib/salestation/web/request_logger.rb (hunk 3 above)
Comment (urmastalimaa) [inline lib/salestation/web/request_logger.rb]: uhh, `Rack body is an array`, not string

Comment (urmastalimaa) [inline lib/salestation/web/request_logger.rb]: maybe `join()` is actually more precise

Comment (urmastalimaa) [inline lib/salestation/web/request_logger.rb]: I read that the iterable structure is used to potentially do some buffering

File: lib/salestation/web/statsd_middleware.rb (hunk 6 above)
Comment (urmastalimaa) [inline lib/salestation/web/statsd_middleware.rb]: https://github.com/salemove/salestation/pull/8/commits/c7213ec108e6dba8bc827c0098d1b655ae2ee38b

File: lib/salestation/web/request_logger.rb (hunk 5 above)
Comment (urmastalimaa) [inline lib/salestation/web/request_logger.rb]: https://github.com/salemove/salestation/pull/8/commits/c7213ec108e6dba8bc827c0098d1b655ae2ee38b

File: lib/salestation/web/request_logger.rb (hunk 3 above)
Comment (urmastalimaa) [inline lib/salestation/web/request_logger.rb]: https://github.com/salemove/salestation/pull/8/commits/c7213ec108e6dba8bc827c0098d1b655ae2ee38b

File: lib/salestation/web/request_logger.rb (hunk 4 above)
Comment (urmastalimaa) [inline lib/salestation/web/request_logger.rb]: https://github.com/salemove/salestation/pull/8/commits/c7213ec108e6dba8bc827c0098d1b655ae2ee38b

File: lib/salestation/web/request_logger.rb (hunk 2 above)
Comment (urmastalimaa) [inline lib/salestation/web/request_logger.rb]: https://github.com/salemove/salestation/pull/8/commits/c7213ec108e6dba8bc827c0098d1b655ae2ee38b

**Comment by urmastalimaa** [comment]
Welp, I now see why `Logger` was stateful. It is so to allow applications to use a logger which outputs `request_id` as metadata.

**Comment by urmastalimaa** [comment]
@indrekj check this https://github.com/salemove/salestation/pull/8/commits/575ff2d7634893b8cd134fc1c7726a65c4f482bd

**Comment by urmastalimaa** [comment]
!merge

**Pull request merged**